base/compiler/tfuncs.jl

```diff
@@ -555,17 +555,22 @@ function typeassert_type_instance(@nospecialize(v), @nospecialize(t))
         widev = widenconst(v)
         if widev <: t
             return v
-        elseif typeintersect(widev, t) === Bottom
+        end
+        tr = typeintersect(widev, t)
+        if tr === Bottom
             return Bottom
         end
-        @assert widev <: Tuple
-        new_fields = Vector{Any}(undef, length(v.fields))
+        @assert tr <: Tuple
```

Comment:
I'm not 100% sure this is safe, as type intersection might be imprecise. Maybe instead substitute `tr = widev` if `!(tr <: Tuple)`?